File: android/app/src/main/java/de/nilu96/retivum/RetivumAppearancePlugin.java
package de.nilu96.retivum;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.Window;
import android.webkit.WebView;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "RetivumAppearance")
public class RetivumAppearancePlugin extends Plugin {
    @PluginMethod
    public void setBackgroundColor(PluginCall call) {
        String value = call.getString("color");
        if (value == null || !value.matches("^#[0-9a-fA-F]{6}$")) {
            call.reject("The supplied background color is invalid", "INVALID_BACKGROUND_COLOR");
            return;
        }
        int color = Color.parseColor(value);

        getBridge().executeOnMainThread(() -> {
            WebView webView = getBridge().getWebView();
            Window window = getActivity().getWindow();
            webView.setBackgroundColor(color);
            window.getDecorView().setBackgroundColor(color);
            window.setBackgroundDrawable(new ColorDrawable(color));
            call.resolve();
        });
    }
}

File: ios/App/App/RetivumAppearancePlugin.swift
import Capacitor
import UIKit

@objc(RetivumAppearancePlugin)
final class RetivumAppearancePlugin: CAPInstancePlugin, CAPBridgedPlugin {
    let identifier = "RetivumAppearancePlugin"
    let jsName = "RetivumAppearance"
    let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setBackgroundColor", returnType: CAPPluginReturnPromise)
    ]

    @objc func setBackgroundColor(_ call: CAPPluginCall) {
        guard let value = call.getString("color"),
              value.range(of: "^#[0-9a-fA-F]{6}$", options: .regularExpression) != nil,
              let color = UIColor.capacitor.color(fromHex: value) else {
            call.reject("The supplied background color is invalid", "INVALID_BACKGROUND_COLOR")
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self, let webView = self.bridge?.webView else {
                call.reject("The native appearance view is unavailable", "APPEARANCE_VIEW_UNAVAILABLE")
                return
            }
            webView.backgroundColor = color
            webView.scrollView.backgroundColor = color
            webView.underPageBackgroundColor = color
            self.bridge?.viewController?.view.backgroundColor = color
            self.bridge?.viewController?.view.window?.backgroundColor = color
            call.resolve()
        }
    }
}
